connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java

```diff
@@ -300,12 +311,105 @@ public ConfigInfos validateConnectorConfig(Map<String, String> connectorProps) {
         } finally {
             Plugins.compareAndSwapLoaders(savedLoader);
         }
     }
 
+    private static ConfigInfos mergeConfigInfos(String connType, ConfigInfos... configInfosList) {
+        int errorCount = 0;
+        List<ConfigInfo> configInfoList = new LinkedList<>();
+        Set<String> groups = new LinkedHashSet<>();
+        for (ConfigInfos configInfos : configInfosList) {
+            if (configInfos != null) {
+                errorCount += configInfos.errorCount();
+                configInfoList.addAll(configInfos.values());
+                groups.addAll(configInfos.groups());
+            }
+        }
+        return new ConfigInfos(connType, errorCount, new ArrayList<>(groups), configInfoList);
+    }
+
+    // public for testing
+    public static ConfigInfos validateClientOverrides(String connName,
+                                                      String prefix,
+                                                      AbstractConfig connectorConfig,
+                                                      ConfigDef configDef,
+                                                      Class<? extends Connector> connectorClass,
+                                                      org.apache.kafka.connect.health.ConnectorType connectorType,
+                                                      ConnectorClientConfigRequest.ClientType clientType,
+                                                      ConnectorClientConfigOverridePolicy connectorClientConfigOverridePolicy) {
+        int errorCount = 0;
+        List<ConfigInfo> configInfoList = new LinkedList<>();
+        Map<String, ConfigKey> configKeys = configDef.configKeys();
+        Set<String> groups = new HashSet<>();
+        Map<String, Object> clientConfigs = connectorConfig.originalsWithPrefix(prefix);
+        for (Map.Entry<String, Object> clientConfig : clientConfigs.entrySet()) {
+            ConfigKey configKey = configKeys.get(clientConfig.getKey());
+            ConfigKeyInfo configKeyInfo = null;
+            if (configKey != null) {
+                if (configKey.group != null) {
+                    groups.add(configKey.group);
+                }
+                configKeyInfo = convertConfigKey(configKey, prefix);
+            }
+            Map<String, Object> clientProps = Collections.singletonMap(clientConfig.getKey(), clientConfig.getValue());
+            ConnectorClientConfigRequest connectorClientConfigRequest = new ConnectorClientConfigRequest(
+                connName, connectorType, connectorClass, clientProps, clientType);
+
+            ConfigValue configValue = new ConfigValue(prefix + clientConfig.getKey(), clientConfig.getValue(),
+                                                      new ArrayList<Object>(), new ArrayList<String>());
+            try {
+                connectorClientConfigOverridePolicy.validate(connectorClientConfigRequest);
```

Comment:
If the passed configs have valid values, we will then not be able to indicate that in the validate response. The only reliable way to indicate what override is invalid is to validate them individually.

Reply:
That was not clear from the KIP or the interface docs, because the request provides a map of values but it's always a singleton. If I were implementing a policy, I would not have expected one property at a time. Plus, this provides no way for the implementation to actually check any combination of values, should that ever be required (we don't use that in the built-in implementations).

I'm not sure what can be done at this point, given the use of PVE to pass the message. It would have been much better to pass in a component with which the implementation could have recorded errors for one or more configuration properties.